Generalize tests for `contlinear`

ericphanson opened the pull request:
I tried starting to address #841. I haven't incorporated @blegat's suggestion to use a type parameter for `TestConfig` but that should be easy to change. I run into a problem, however, when trying to run the `contlinear` tests with `BigFloat`s for SDPA-GMP, which is that `Float64`'s are used internally in a few places in MOI. For example,

https://github.com/JuliaOpt/MathOptInterface.jl/blob/6ac4e6d6ee47486062ab9a14816aca6d51328760/src/Bridges/Constraint/slack.jl#L20

causes
```julia
  MethodError: no method matching shift_constant(::MathOptInterface.EqualTo{Float64}, ::BigFloat)
```
in the test on line `1539` of `contlinear.jl`. I imagine this is why it's important to test with other numeric types if we want to use them with MOI! I think therefore something beyond changing the tests here is needed. Perhaps auditing the codebase for `Float64` usage and replacing it with an appropriate type? It looks easy to fix in the slack.jl#L20 case. I saw also `Float64`'s here:

https://github.com/JuliaOpt/MathOptInterface.jl/blob/6ac4e6d6ee47486062ab9a14816aca6d51328760/src/Bridges/bridge_optimizer.jl#L724

and I'm not sure what the replacement would be, because `SingleVariable`'s don't carry a type parameter.

Regarding the changes in this PR:

My technique for updating the file was lots of find-and-replace. There are four types of constructors which need to be parametrized (VAFs, SAFs, SATs, and VATs), and integers written as floating point numbers (`3.0`) can be regex replaced by finding `(\d*?_?\d+?)\.0` and replacing by `T($1)` (assuming `T` is the numeric type). I also first replaced `0.0`'s by `zero(T)` and `1.0` by `one(T)`, although I'm not sure if that's better than `T(0)` and `T(1)`. Then I searched for remaining decimal points and replaced them with fractions by hand (e.g. `T(1)/T(2)`).

I'm not sure if replacing the constant literals was desired; @odow suggested just checking that they promote correctly. But because I was running into lots of promotion-related errors when running the `contlinear` test with SDPA-GMP, I decided to replace them. I think that helped isolate which problems were coming from MOI's internal use of Float64 numbers, versus ones introduced in `contlinear.jl`.

ericphanson commented on src/Test/config.jl:
```diff
@@ -10,13 +10,14 @@ struct TestConfig
     # MOI.OPTIMAL or MOI.LOCALLY_SOLVED.
     optimal_status::MOI.TerminationStatusCode
     basis::Bool # can get variable and constraint basis status
+    number_type::Type{<:Number}
```
Parametrizing TestConfig would get rid of the abstract type here, which is probably desirable

ericphanson commented:
I parametrized `TestConfig`. I have two questions:

1. Should we use `one(T)` and `zero(T)` or `T(1)` and `T(0)`? If `one(T)` is really better, what about `100*one(T)` vs `T(100)`? I think in the `one(T)` version, we are allowing a custom numeric type to have some kind of strange multiplicative identity which is not `==1.0`, whereas in the `T(1)` picture we are assuming that we want something essentially `==1.0` and that we can construct our type from integers without loss of precision.
2. How do we test this? It seems like maybe we need to add tests that use this code with various numeric types. And do we need to add tests for the `TestConfig` object itself?

ericphanson commented on src/Test/contlinear.jl:
```diff
@@ -333,12 +333,12 @@ function linear1test(model::MOI.ModelLike, config::TestConfig)
-            @test MOI.get(model, MOI.ConstraintDual(), c) ≈ -1.5 atol=atol rtol=rtol
-            @test MOI.get(model, MOI.ConstraintDual(), c2) ≈ 0.5 atol=atol rtol=rtol
+            @test MOI.get(model, MOI.ConstraintDual(), c) ≈ -T(3)/T(2) atol=atol rtol=rtol
+            @test MOI.get(model, MOI.ConstraintDual(), c2) ≈ T(1)/T(2) atol=atol rtol=rtol
```
Ah that does sound better, I’ll change those

blegat commented on src/Test/config.jl:
```diff
@@ -10,14 +10,15 @@ struct TestConfig
+        basis::Bool = false) where {T <: Number}
         new(atol, rtol, solve, query, modify_lhs, duals, infeas_certificates,
-            optimal_status, basis)
+            optimal_status, basis) 
```
Remove trailing space

blegat commented:
@odow `Complex{Rational}` is not a concrete type hence the inference failure. With `Complex{Rational{Int}}`, both give the same thing:
```julia
julia> f(x) = x * one(Complex{Rational{Int}})
f (generic function with 2 methods)

julia> g(x) = Complex{Rational{Int}}(x)
g (generic function with 2 methods)

julia> @code_warntype g(1)
Variables
  #self#::Core.Compiler.Const(g, false)
  x::Int64

Body::Complex{Rational{Int64}}
1 ─ %1 = Core.apply_type(Main.Rational, Main.Int)::Core.Compiler.Const(Rational{Int64}, false)
│   %2 = Core.apply_type(Main.Complex, %1)::Core.Compiler.Const(Complex{Rational{Int64}}, false)
│   %3 = (%2)(x)::Complex{Rational{Int64}}
└──      return %3

julia> @code_warntype g(1)
Variables
  #self#::Core.Compiler.Const(g, false)
  x::Int64

Body::Complex{Rational{Int64}}
1 ─ %1 = Core.apply_type(Main.Rational, Main.Int)::Core.Compiler.Const(Rational{Int64}, false)
│   %2 = Core.apply_type(Main.Complex, %1)::Core.Compiler.Const(Complex{Rational{Int64}}, false)
│   %3 = (%2)(x)::Complex{Rational{Int64}}
└──      return %3
```
I prefer `zero(T)` and `one(T)` over `T(0)` and `T(1)` since `one` and `zero` are more specific than `T(::Int)`. However, for other number `x`, `T(x)` seems better than `x * one(T)` or `x + zero(T)`.

blegat commented:
https://github.com/JuliaOpt/MathOptInterface.jl/blob/6ac4e6d6ee47486062ab9a14816aca6d51328760/src/Bridges/Constraint/slack.jl#L20
Should be `zero(T)`.
https://github.com/JuliaOpt/MathOptInterface.jl/blob/6ac4e6d6ee47486062ab9a14816aca6d51328760/src/Bridges/bridge_optimizer.jl#L724
That's trickier.
One solution would be to create a new function in the MOI.AbstractBridgeOptimizer API to get the MOI.Bridges.Constraint.ScalarFunctionizeBridge.
For `LazyBridgeOptimizer`, it will loop through the list of of bridges and return the first one with this type.

ericphanson commented:
> I prefer `zero(T)` and `one(T)` over `T(0)` and `T(1)` since `one` and `zero` are more specific than `T(::Int)`. However, for other numbers, say `x`, I agree that `T(x)` seems better than `x * one(T)` or `x + zero(T)`.

Ok, I reverted that to go back to `zero(T)` and `one(T)`.

So in summary (and to make it easy to do these changes for other tests):
* `one(T)` and `zero(T)` to replace `0.0` and `1.0`
* `T(n)` to replace `n.0` (integer `n`)
* `T(n//m)` to replace floats like `1.5`

Besides that, I added type parameters to `MOI.EqualTo`, `MOI.GreaterThan`, `MOI.LessThan`, and `MOI.ScalarAffineTerm`, `MOI.ScalarAffineFunction`, `MOI.VectorAffineTerm`, and `MOI.`VectorAffineFunction`.

To test the changes (which helped me find that `MOI.EqualTo` and the other similar constraints needed to be parametrized), I've been using the script
```
using Test

using MathOptInterface
const MOI = MathOptInterface
const MOIT = MOI.Test
const MOIU = MOI.Utilities
const MOIB = MOI.Bridges

import SDPA_GMP

function test_with_type(T)
    optimizer = SDPA_GMP.Optimizer{T}()
    MOI.set(optimizer, MOI.Silent(), true)

    @testset "SolverName" begin
        @test MOI.get(optimizer, MOI.SolverName()) == "SDPA-GMP"
    end

    @testset "supports_default_copy_to" begin
        @test MOIU.supports_allocate_load(optimizer, false)
        @test !MOIU.supports_allocate_load(optimizer, true)
    end

    # UniversalFallback is needed for starting values, even if they are ignored by SDPA
    cache = MOIU.UniversalFallback(MOIU.Model{T}())
    cached = MOIU.CachingOptimizer(cache, optimizer)
    bridged = MOIB.full_bridge_optimizer(cached, T)
    # test 1e-3 because of rsoc3 test, otherwise, 1e-5 is enough
    config = MOIT.TestConfig{T}(atol=1e-3, rtol=1e-3)

    @testset "MOIT.contlineartest" begin
        MOIB.remove_bridge(bridged, MOIB.Constraint.ScalarSlackBridge{T})
        MOIT.contlineartest(bridged, config, [
            "linear8a",
            "linear12"
        ])
    end

end

@testset "Tests with type $T" for T in (Float64, BigFloat)
    test_with_type(T)
end
```
It still has some errors on `BigFloat` runs, but I think those are due to the use of `Float64` in MOI itself.

blegat commented on src/Test/contlinear.jl:
```diff
@@ -577,42 +577,42 @@ function linear4test(model::MOI.ModelLike, config::TestConfig)
-    MOI.set(model, MOI.ConstraintSet(), c2, MOI.LessThan(-100.0))
+    # s.t. T(100) <= x
+    #               y <= -T(100)
+    MOI.set(model, MOI.ConstraintSet(), c2, MOI.LessThan{T}(-T(100)))
```
`MOI.LessThan(T(-100))` should be sufficient, no need to do `MOI.LessThan{T}`

blegat commented:
> which helped me find that MOI.EqualTo and the other similar constraints needed to be parametrized

Why is that ? It automatically picks the type of the argument:
```julia
julia> MOI.EqualTo(1)
MathOptInterface.EqualTo{Int64}(1)
```

ericphanson commented:
> > which helped me find that MOI.EqualTo and the other similar constraints needed to be parametrized
> 
> Why is that ? It automatically picks the type of the argument:
> 
> ```julia
> julia> MOI.EqualTo(1)
> MathOptInterface.EqualTo{Int64}(1)
> ```

Hmm, well it might be simply because after I parametrized the functions, things like:

https://github.com/JuliaOpt/MathOptInterface.jl/blob/111fba5622b1e8ce1055dce2e9f01b2ad14ad971/src/Test/contlinear.jl#L15

won't work unless I also change the constraint there. I didn't realize that I only needed to add the type parameter in some places though, and just find-and-replaced it everywhere.

chriscoey commented:
We may as well make obvious style fixes while making these changes. There should be spaces between elements of vectors, like `[one(T), zero(T)]` rather than `[one(T),zero(T)]`

blegat commented:
@ericphanson Indeed, it's needed when you just write the type but not when it's a constructor

ericphanson commented:
@blegat Ok, I changed it to only put type parameters when it is not used as a constructor. I'll update my little summary of changes.

ericphanson commented:
> We may as well make obvious style fixes while making these changes. There should be spaces between elements of vectors, like `[one(T), zero(T)]` rather than `[one(T),zero(T)]`

Ok, I made that specific fix (regex-find `(\[.+?),([^ ])`, replace `$1, $2`)

blegat commented on src/Test/contlinear.jl:
```diff
@@ -1,7 +1,7 @@
 
 # Basic solver, query, resolve
-function linear1test(model::MOI.ModelLike, config::TestConfig)
+function linear1test(model::MOI.ModelLike, config::TestConfig{T}) where {T <: Number}
```
Why adding `<: Number` here ? As `T` is a type param of `TestConfig`, it is already constrained to be `<: Number`

ericphanson replied:
No big reason; I thought maybe it adds clarity to the reader about what T is for, but I agree it’s not needed. Should I take these off?

blegat replied:
Yes, it's more consistent with the rest of the MOI code

ericphanson replied:
Ok, will do
